### PR TITLE
3046015 path patterns

### PR DESCRIPTION
**Pull request by safalliajoseph**
d.o Issue: https://www.drupal.org/project/3034968/issues/3046015

### CODE REVIEW

**Comment by safalliajoseph** [comment]
@krystalcode I created a new MR for path pattern configs as the old branch was having history related issues. I have corrected the feedbacks you gave in https://github.com/shabananavas/elastic_apm/pull/21#issuecomment-533387432.

File: src/EventSubscriber/RequestSubscriber.php [hunk 1]
```diff
@@ -271,12 +283,63 @@ protected function constructQuerySpan($connection, $driver, array $query) {
...
+   *
+   * @return array
+   *   An array of tags to pass to the PHP Agent.
+   */
+  protected function preparePathPatternTags() {
+    $tags = [];
+
+    // Fetch the current path from route object.
+    $route_object = $this->routeMatch->getRouteObject();
+    $path = $route_object->getPath();
```
Comment (krystalcode) [inline src/EventSubscriber/RequestSubscriber.php]: **CHANGE REQUEST**

No need for intermediate unused variables, habitually doing this increases memory consumption for no reason:

```
$path = $this->routeMatch->getRouteObject()->getPath();
```

File: src/EventSubscriber/RequestSubscriber.php [hunk 2]
```diff
@@ -271,12 +283,63 @@ protected function constructQuerySpan($connection, $driver, array $query) {
...
+
+    // Fetch the configured path patterns.
+    $tag_config = $this->apiService->getTagConfig();
+    $path_pattern = $tag_config['path_patterns'];
```
Comment (krystalcode) [inline src/EventSubscriber/RequestSubscriber.php]: **CHANGE REQUEST**

Same here:

```
$path_pattern = $this->apiService->getTagConfig()['path_patterns'];
```

File: src/EventSubscriber/RequestSubscriber.php [hunk 3]
```diff
@@ -271,12 +283,63 @@ protected function constructQuerySpan($connection, $driver, array $query) {
...
+
+    if (empty($path_pattern)) {
+      return;
+    }
```
Comment (krystalcode) [inline src/EventSubscriber/RequestSubscriber.php:316]: **CHANGE REQUEST**

If the configuration might not contain the `path_pattern` key at all, then you'd get a PHP warning here and you need to to check if the key exists before using it:

```
$tag_config = $this->apiService->getTagConfig();
if (empty($tag_config['path_pattern'])) {
  return;
}
$path_pattern = $tag_config['path_patterns'];
```

If the key will always exist, you don't need to add the overhead of `empty`, or even `isset`:

```
$path_pattern = $this->apiService->getTagConfig()['path_patterns'];
if (!$path_pattern) {
  return;
}
```

File: src/EventSubscriber/RequestSubscriber.php [hunk 4]
```diff
@@ -271,12 +283,63 @@ protected function constructQuerySpan($connection, $driver, array $query) {
...
+
+    // Explode path patterns by new line.
+    $path_patterns = explode(PHP_EOL, $path_pattern);
+
+    // Add tags depending on the path pattern set.
+    foreach ($path_patterns as $path_pattern) {
```
Comment (krystalcode) [inline src/EventSubscriber/RequestSubscriber.php]: **CHANGE REQUEST**

It's not good practice in terms of writing clear code and code that is not but-prone to reuse names for unrelated variables. `$path_pattern` defined above is different than the `$path_pattern` within the loop, name it differently: `foreach ($path_patterns as $pattern) {`.

File: src/EventSubscriber/RequestSubscriber.php [hunk 5]
```diff
@@ -271,12 +283,63 @@ protected function constructQuerySpan($connection, $driver, array $query) {
...
+      $patterns = explode(':', $path_pattern);
+
+      // If the configured path does not match with the current path
+      // continue to look for the next pattern.
+      if (!($this->pathMatcher->matchPath($path, $patterns['0']))) {
```
Comment (krystalcode) [inline src/EventSubscriber/RequestSubscriber.php]: **CHANGE REQUEST**

`if (!$this->pathMatcher->matchPath($path, $patterns['0'])) {`

File: src/Form/TagSettingsForm.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        Example:
+        <pre>
+          /product/*: provider|commerce
+          /checkout/*: provider|commerce
+        </pre>
+      '),
+    ];
+
+    return parent::buildForm($form, $form_state);
+  }
```
Comment (krystalcode) [inline src/Form/TagSettingsForm.php:88]: **CHANGE REQUEST**

Add some validation. Make sure that every pattern is composed of the 3 expected parts.

File: src/EventSubscriber/RequestSubscriber.php [hunk 7]
```diff
@@ -271,12 +283,63 @@ protected function constructQuerySpan($connection, $driver, array $query) {
...
+        continue;
+      }
+
+      // Do not proceed if a tag value is not set.
+      if (empty($patterns['1'])) {
+        continue;
+      }
```
Comment (krystalcode) [inline src/EventSubscriber/RequestSubscriber.php]: **CHANGE REQUEST**

Should not be needed if we add validation in the form.

File: src/EventSubscriber/RequestSubscriber.php [hunk 8]
```diff
@@ -271,12 +283,63 @@ protected function constructQuerySpan($connection, $driver, array $query) {
...
+    if (empty($path_pattern)) {
+      return;
+    }
+
+    // Explode path patterns by new line.
+    $path_patterns = explode(PHP_EOL, $path_pattern);
+
+    // Add tags depending on the path pattern set.
+    foreach ($path_patterns as $path_pattern) {
+      $patterns = explode(':', $path_pattern);
```
Comment (krystalcode) [inline src/EventSubscriber/RequestSubscriber.php]: **CHANGE REQUEST**

Now this is where it all becomes confusing and bug-prone with variable names. What is the difference between `$path_pattern`, $path_patterns`, `$patterns`?

We need to do some refactoring here. Create a function that takes the configuration value and returns an array of already parsed pattern/tag pairs:

```
[
  [
    'pattern' => '/product/*',
    'tag_key' => 'provider',
    'tag_value' => 'commerce',
  ],
  [
    'pattern' => '/checkout/*',
    'tag_key' => 'provider',
    'tag_value' => 'commerce',
  ],
];
```

So:

```
$patterns = $this->parsePatterns($this->apiService->getTagConfig());
foreach ($patterns as $pattern) {
  if (!$this->pathMatcher->matchPath($path, $patterns['pattern'])) {
    continue;
  }

  $tags[$pattern['tag_key']] = $pattern['tag_value'];
}
```
The above will result in more readable and concise code without variable naming problems. You will also be able to reuse part of the code for processing route patterns as well.

**Review by krystalcode — CHANGES_REQUESTED** [review]

File: src/EventSubscriber/RequestSubscriber.php (hunk 1 above)
Comment (safalliajoseph) [reply src/EventSubscriber/RequestSubscriber.php]: Done.

File: src/Form/TagSettingsForm.php (hunk 6 above)
Comment (safalliajoseph) [reply src/Form/TagSettingsForm.php:88]: Done. Added validation

File: src/EventSubscriber/RequestSubscriber.php (hunk 7 above)
Comment (safalliajoseph) [reply src/EventSubscriber/RequestSubscriber.php]: Yep right. Removed

File: src/EventSubscriber/RequestSubscriber.php (hunk 8 above)
Comment (safalliajoseph) [reply src/EventSubscriber/RequestSubscriber.php]: Added a new method `parsePatterns` in apiService class

**Comment by safalliajoseph** [comment]
@krystalcode I have corrected the issues you noticed. Could you please review again. Thanks!

**Comment by krystalcode** [comment]
Merged on Drupal.org

**Pull request closed without merging**